WordPress/gutenberg
# Release tool: warn about open issues in milestone before starting release

Closes #20151

This pull request seeks to enhance the release tool to check whether there are open issues or pull requests in the milestone associated with the version being released. #20151 includes a detailed of use-cases relevant for this behavior:

>As part of the release process, it is typically recommended to look at the pending release milestone to assure there are no lingering efforts in progress. [...]
>- The release coordinator is not familiar with this recommendation, and proceeds with publishing the release without checking the milestone, thus potentially missing some important changes to have been included.
>- For last-minute blockers, a developer could include an issue or pull request in a milestone to flag a release as being ineligible, in case the release coordinator misses pings in said pull requests, issues, or in Slack.

This mostly refactors and reuses existing code which had been implemented for the changelog tool, which has similar requirement to query issues (pull requests) for a given milestone.

Currently, the check occurs after the clone step of the release, since it's the earliest point that the intended release version is fully known. Ideally it could happen earlier, though it may require additional refactoring to the release tool to determine this version earlier.

**Testing Instructions:**

Ensure unit tests pass:

```
npm run test-unit bin/plugin/lib/test/milestone.js
```

It could be tested by starting the release and aborting it earlier.

Alternatively, I tested this by temporarily adding `isMilestoneClear` as a `module.exports` of the `release.js` file, then using `node` REPL to confirm expected behavior:

```
⇒ node
Welcome to Node.js v12.13.0.
Type ".help" for more information.
> const { isMilestoneClear } = require( './bin/plugin/commands/release' );
undefined
> isMilestoneClear( '8.2.0' ).then( ( isClear ) => console.log( isClear ? 'Clear' : 'Nope' ) );
Promise { <pending> }
> Nope
> isMilestoneClear( '8.3.0' ).then( ( isClear ) => console.log( isClear ? 'Clear' : 'Nope' ) );
Promise { <pending> }
> Clear
```

Ensure there are no regressions in generating changelog entries:

```
npm run changelog
```
